towhee/functional/mixins/parallel.py

```diff
@@ -189,54 +305,143 @@ def mmap(self, *arg):
-                for i in range(len(arg)):
+                for i in range(len(ops)):
                     queue = queues[i]
                     buff = buffs[i]
                     queue.put(await buff.pop(0))
-            nonlocal flag
-            flag = False
+            for i in range(len(ops)):
+                queue = queues[i]
+                poison = _PoisonPill()
+                queue.put(poison)
 
         def worker_wrapper():
             loop.run_until_complete(worker())
 
-        executor = self.get_executor()
-        if executor is None:
-            executor = concurrent.futures.ThreadPoolExecutor(len(arg))
-        num_worker = 1
-        queues = [Queue(maxsize=num_worker) for _ in arg]
+        t = threading.Thread(target=worker_wrapper)
+        t.start()
+
+        retval = [inner(queue) for queue in queues]
+        return [self.factory(x) for x in retval]
+
+    def _ray_mmap(self, ops, num_worker=None):
+        import ray #pylint: disable=import-outside-toplevel
+
+        if num_worker is not None:
+            pass
+        elif self.get_executor() is not None:
+            num_worker = self._num_worker
+        else:
+            num_worker = 2
+
+        queues = [Queue(maxsize= max(1, num_worker//len(ops) + 1)) for _ in ops]
         loop = asyncio.new_event_loop()
-        flag = True
 
-        executor.submit(worker_wrapper)
+        def inner(queue):
+            flag = True
+            while flag:
+                x = queue.get()
+                if isinstance(x, _PoisonPill):
+                    flag = False
+                else:
+                    yield x
+
+        async def worker():
+            buffs = [[] for _ in ops]
+            for x in self:
+                for i in range(len(ops)):
+
+                    @ray.remote
+                    def remote_runner(val):
+                        #TODO: double check working correctly
+                        return _map_task_ray(ops[i])(val) #pylint: disable=cell-var-from-loop
+
+                    queue = queues[i]
+                    buff = buffs[i]
+                    if len(buff) == num_worker:
+                        queue.put(await buff.pop(0))
+                    buff.append(
+                        asyncio.wrap_future(remote_runner.remote(x).future()))
+
+            while sum([len(buff) for buff in buffs]) > 0:
+                for i in range(len(ops)):
+                    queue = queues[i]
+                    buff = buffs[i]
+                    queue.put(await buff.pop(0))
+
+            for i in range(len(ops)):
+                queue = queues[i]
+                poison = _PoisonPill()
+                queue.put(poison)
+
+
+        def worker_wrapper():
+            loop.run_until_complete(worker())
+
+        t = threading.Thread(target=worker_wrapper)
+        t.start()
 
         retval = [inner(queue) for queue in queues]
         return [self.factory(x) for x in retval]
 
+class _PoisonPill():
```

Comment:
The name `PoisonPill` makes people confused, EOS (end of stream, or EO-X, end of something) might be a better name.